text/0000-context-service.md

```diff
@@ -0,0 +1,200 @@
+
+Invariants:
+
+* `setContextExperimental()` must be called at least once before the insertion into the DOM phase is completed.
+
+Observations:
+* LWC is not opinionated on how a context definition is going to be share. User-land abstractions are allowed.
+* LWC is not opinionated on how a developer will share the capabilities to set a new context value on an element. User-land abstractions are allowed.
+
+## Other examples
+
+### How to implement a React-like context HOC in LWC?
+
+```js
+import { LightningElement, createContextExperimental, setContextExperimental } from 'lwc';
+
+// creating a new context definition
+const Context = createContextExperimental({
+    x: 0,
+    y: 0,
+});
+
+export default class XFooContext extends LightningElement {
+    _value = {};
+
+    @api
+    get theme() {
+        return this._value;
+    }
+    set theme(v) {
+        // you can validate `v` here if you want
+        this._value = v;
+        // setting a new context value 
+        this.setNewContextValue();
+    }
+
+    connectedCallback() {
+        // setting the initial value of the context
+        this.setNewContextValue();
+    }
+    setNewContextValue() {
+        setContextExperimental(this, Context, this._value);
+    }
+    static Consumer = Context;
+}
+```
+
+The way a provider is used in a template is very straight forward:
+
+```html
+<template>
+    <x-foo-context theme="dark">
+        <c-consumer></c-consumer> 
+    </x-foo-context>
+    <x-foo-context value="lite">
+        <c-other-consumer></c-other-consumer> 
+    </x-foo-context>    
+</template>
+```
+
+Where `<c-consumer>` and `<c-other-consumer>` or any of their descendent will be able to consume the context value by using the `@wire(XFooContext.Consumer)` decorator, e.g.:
+
+```js
+import { LightningElement } from 'lwc';
+import XFooContext from 'x/fooContext';
+
+export default class MyComponent extends LightningElement {
+    @wire(XFooContext.Consumer) foo; // context object provided by <x-foo-context> element
+}
+```
+
+# Other Alternatives
+
+* The two new APIs can be imported from `@lwc/context` or `@lwc/wire-service` rather than `lwc` directly. This might help to eventually allow others to use this contexts with vanilla web components. It will also make it easier for us to restrict who has access to these APIs in platform.
+* Make the new APIs a lot more generic, instead of context specific, where you might want to use them to provide contextual wiring per adapters. The fear here is that we will give them a very powerful experimental API.
+
+# Adoption strategy
+
+This is a brand new feature, we just need to document it.
+
+# How we teach this
+
+* The consumer aspect of it is very straight forward considering that we have plenty of documentation about the `@wire` decorator.
+
+# Unresolved questions
```

Comment:
A couple more:
- Is the value always synchronous provisioned during the `wiring` stage of the component lifecycle?  
- Is there a formal error state that the wire can provision? Will it participate in the component error boundary?

Reply:
added